chainer/functions/activation/log_softmax.py

```diff
@@ -0,0 +1,111 @@
+import numpy
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+if cuda.cudnn_enabled:
+    cudnn = cuda.cudnn
+    libcudnn = cudnn.cudnn
+    _algorithm = libcudnn.CUDNN_SOFTMAX_LOG
+    _mode = libcudnn.CUDNN_SOFTMAX_MODE_CHANNEL
+    _cudnn_version = libcudnn.getVersion()
+
+
+def logsumexp(x):
+    xp = cuda.get_array_module(x)
+    m = x.max(axis=1, keepdims=True)
+    y = x - m
+    xp.exp(y, out=y)
+    return xp.log(y.sum(axis=1, keepdims=True)) + m
+
+
+class LogSoftmax(function.Function):
+
+    """Log-softmax activation function."""
+
+    def __init__(self, use_cudnn=True):
+        self.use_cudnn = use_cudnn
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        x_type, = in_types
+
+        type_check.expect(
+            x_type.dtype == numpy.float32,
+            x_type.ndim > 1,
+        )
+
+    def forward(self, xs):
+        x = xs[0]
+        xp = cuda.get_array_module(x)
+        if xp != numpy and cuda.cudnn_enabled and self.use_cudnn \
+           and _cudnn_version >= 3000:
+            dtype = x.dtype
+            one = numpy.array(1, dtype=dtype).ctypes
+            zero = numpy.array(0, dtype=dtype).ctypes
+            handle = cudnn.get_handle()
+            x_cube = x.reshape(x.shape[:2] + (-1, 1))
+            desc = cudnn.create_tensor_descriptor(x_cube)
+            self.y = xp.empty_like(x)
+            libcudnn.softmaxForward(
+                handle, _algorithm, _mode, one.data, desc.value,
+                x_cube.data.ptr, zero.data, desc.value,
+                self.y.data.ptr)
+            return self.y,
+
+        else:
+            log_z = logsumexp(x)
+            self.y = x - log_z
+            return self.y,
+
+    def backward(self, x, gy):
+        xp = cuda.get_array_module(*x)
+        if xp != numpy and cuda.cudnn_enabled and self.use_cudnn \
+           and _cudnn_version >= 3000:
+            dtype = x[0].dtype
+            one = numpy.array(1, dtype=dtype).ctypes
+            zero = numpy.array(0, dtype=dtype).ctypes
+            handle = cudnn.get_handle()
```

Comment:
If we can assume `backward` is always preceded by `forward`, we can store `one` and `zero` in `forward` and use them in `backward`.